util.c

```diff
@@ -1145,3 +1149,116 @@ void drop_cookie_secret(struct nsd* nsd)
 	              , NSD_COOKIE_SECRET_SIZE);
 	nsd->cookie_count -= 1;
 }
+
+static
+int cookie_secret_file_read(nsd_type* nsd) {
+	cookie_secret_type cookie_secrets[NSD_COOKIE_HISTORY_SIZE];
+	char secret[NSD_COOKIE_SECRET_SIZE * 2 + 2/*'\n' and '\0'*/];
+	FILE* f;
+	size_t count = 0;
+	const char* fn;
+
+	if(!(fn = cookie_secret_file(nsd->options)))
+		return 1; /* Explicitely disabled with empty filename */
+
+	else if((f = fopen(fn, "r")) != NULL)
+		; /* pass */
+
+	/* a non-existing cookie file is not an error */
+	else if(errno != ENOENT) {
+		log_msg( LOG_ERR
+		       , "error reading cookie secret file \"%s\": \"%s\""
+		       , fn, strerror(errno));
+		return 0;
+	}
+	else if(nsd->options->cookie_secret_file != NULL /* explicit name */
+	     ||!(f = fopen((fn = CONFIGDIR"/nsd_cookiesecrets.txt"), "r")))
+		return 1;
+
+	/* cookie secret file exists and is readable */
+	for( count = 0; count < NSD_COOKIE_HISTORY_SIZE; count++ ) {
+		size_t secret_len = 0;
+		ssize_t decoded_len = 0;
+		if( fgets(secret, sizeof(secret), f) == NULL ) { break; }
+		secret_len = strlen(secret);
+		if( secret_len == 0 ) { break; }
+		assert( secret_len <= sizeof(secret) );
+		secret_len = secret[secret_len - 1] == '\n' ? secret_len - 1 : secret_len;
+		if( secret_len != NSD_COOKIE_SECRET_SIZE * 2 ) {
+			fclose(f);
+			log_msg( LOG_ERR
+			       , "error parsing cookie secret file \"%s\""
+			       , fn);
+			explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
+			explicit_bzero(secret, sizeof(secret));
+			return 0;
+		}
+		/* needed for `hex_pton`; stripping potential `\n` */
+		secret[secret_len] = '\0';
+		decoded_len = hex_pton(secret, cookie_secrets[count].cookie_secret,
+		                       NSD_COOKIE_SECRET_SIZE);
+		if( decoded_len != NSD_COOKIE_SECRET_SIZE ) {
+			fclose(f);
+			log_msg( LOG_ERR
+			       , "error parsing cookie secret file \"%s\""
+			       , fn);
+			explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
+			explicit_bzero(secret, sizeof(secret));
+			return 0;
+		explicit_bzero(secret, sizeof(secret));
+		}
+	}
+	fclose(f);
+	if(count && nsd->cookie_secrets_source <= COOKIE_SECRETS_FROM_FILE) {
```

Comment:
An `enum` is mapped to an integer, but it's better to be explicit. The order might change and this may go unnoticed. Basically, you want the branch to be taken if the secret does not originate from the configuration (`!= COOKIE_SECRETS_FROM_CONFIG`).

Reply:
Done in [5674228](https://github.com/NLnetLabs/nsd/pull/378/commits/5674228f1cd200bbf1e5a645d55e8bbf0b90dd3d)